### PR TITLE
Application: mining node

### PR DESCRIPTION
**Pull request by dunnock**
## Description
This PR adds mining node application. Mining node has minimum dependency on tari libraries as to make it good example for portability (so far it is still dependent on tari_core BlockHeader and ProofOfWork math), further decoupling might be implemented in follow up PRs. To get the data necessary for mining node requires grpc connection to tari base node and console wallet.

Mining is performed via stages:
1. Validate if base node is bootstrapped
2. Retrieve block template from base node
3. Retrieve coinbase from console wallet
4. Retrieve new block from base node based on template enriched with coinbase output and kernel
5. Mine for block with matching difficulty
6. Submit mined block to base node and restart from 1

When started locally mining node will try to get settings from `[mining_node]` section, including `base_node_grpc_address` and `wallet_grpc_address`. If grpc addresses not set it will try to load from relevant sections from a global config, so when both wallet and node are runnign locally it can start with zero-configuration. By default if `num_mining_threads` setting is not set it will use number of logical CPU cores. It also supports network section overrides.

In addition PR also fixes ConfigLoader trait to use `network` setting (it was still using `use_network` which I think was changed some time ago). Also changes some usize numbers wrapping for current CpuMiner implementation to rust native.

Also there is a little improvement in initialization of applications, as all applications seems be doing same - initialize bootstrap, initialize logs and load global config moved this part to one library function for base node, wallet and mining node.

I've also started moving sha3 calculation into mining node, though left that not completed as it is being changed now.

## Motivation and Context
This application would give example of how to write a miner node for tari, which might spike more opensource miners implementation.

## How Has This Been Tested?
I have run mining node in local network with 2 nodes, one of nodes was mining too. Fixed tests for config changes for config loaders.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [X] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [X] I'm merging against the `development` branch.
* [X] I ran `cargo-fmt --all` before pushing.
* [X] I ran `cargo test` successfully before submitting my PR.
* [X] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: applications/tari_app_utilities/src/initialization.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+use crate::utilities::ExitCodes;
```
Comment (delta1) [inline applications/tari_app_utilities/src/initialization.rs:1]: new files should have the copyright preamble I think

File: applications/tari_mining_node/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
+//! Miner specific configuration
```
Comment (delta1) [inline applications/tari_mining_node/src/config.rs:23]: copyright

File: applications/tari_mining_node/src/difficulty.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
+use sha3::{Digest, Sha3_256};
```
Comment (delta1) [inline applications/tari_mining_node/src/difficulty.rs:23]: copyright

**Review by delta1 — CHANGES_REQUESTED** [review]
Looking good @dunnock ! 

A few things, I know this is wip: 

- new files need the copyright/licence preamble at the top 
- tari_mining_node crate needs a readme with some instructions 
- maybe `use log::*` to bring log functions into scope instead of having to them all fully qualified 
- config options should be listed in the config samples as well (eg: windows.toml) 

I'm still trying to get things working enough to actually test the mining! But that's due to network and node syncing issues 

Personally I think that the crate could be named `tari_miner` - what do you think @mikethetike ?

**Comment by dunnock** [comment]
Thank you for feedback @delta1 , added copyrights, README and config examples. Also imported `log::` stuff. Retained from renaming until Mike would confirm.

**Comment by dunnock** [comment]
I'm also waiting for https://github.com/tari-project/tari/pull/2514 to land to accomodare grpc changes in this PR.

**Review by stringhandler — COMMENTED** [review]
Looks good, will test locally

**Pull request merged**